### DISCUSSION TITLE
Python’s Preprocessor

### DISCUSSION THREAD

**Discussion by giscus** [pydongery/blog#5] [Blog]
# Python’s Preprocessor

Every now and the you hear outrageous claims such as “Python has no preprocessor”.

https://pydong.org/posts/PythonsPreprocessor/

**Comment by sdmagic** [comment]
Two things you missed in your Preprocessor talk (which was very good!):

### Constants
* In C/C++, they're much more than just naming conventions
* The source code can't modify constants
* The constant value is plugged into your code by the preprocessor, then the constant name is thrown away.

### Macros
* They're basically constants on steroids
* Can contain arguments that are expanded by the constant

**Comment by etiennebonnafoux** [comment]
Great post! Thanks for it.

**Comment by pirate** [comment]
This is one of the more wild things I've seen in 15 years of python coding. Great post!

Definitely saving this for some future nightmare scenario where I might need it.

**Comment by awesomebytes** [comment]
Thanks for this magnific post. I started looking into how `from __future__` stuff was implemented but I got a bit overwhelmed. With your post I understand way better how it can be done.
I also love hacky stuff like this, even more if it allows to do "impossible" cross-language tricks.

**Comment by cfbolz** [comment]
What an amazing post, thank you! I particularly love the cppyy example (and cppyy an amazing library!). 

I'll send this to everyone from now on who tells me that doing static analysis of python 'should be easy' (or static compilation).

**Comment by auscompgeek** [comment]
More example usages:
- backport of f-strings to Python < 3.6: https://github.com/asottile-archive/future-fstrings
- backport of `from __future__ import annotations` to Python < 3.7: https://github.com/asottile-archive/future-annotations

**Reply by Tsche** [reply]
Someone on HN also mentioned https://github.com/dropbox/pyxl this very impressive real world example :)
